.changeset/tidy-oranges-allow.md

```diff
@@ -0,0 +1,12 @@
+---
+'@backstage/plugin-scaffolder-backend': minor
+---
+
+Added experimental flag for scaffolder to wait for running tasks to complete on shutdown
+
+Enabling the `EXPERIMENTAL_gracefulShutdown` flag in the scaffolder config will make the
+scaffolder block the shutdown process until all running tasks have completed. This is useful
+when there is a need to ensure that all tasks have completed before the scaffolder is shut down.
+
+Please note, that the `TaskWorker` `stop` method is now asynchronous and awaited for the
+tasks to complete when the experimental flag is enabled.
```

Comment:
Hey, did you test it on k8s redeploy scenario?

Reply:
This should work pretty nicely with the new liveness and readiness checks with the new backend system, it's what I was gonna test internally too.

https://github.com/backstage/backstage/pull/27961

[these docs](https://backstage.io/docs/backend-system/core-services/root-http-router/#via-app-configyaml) might also be super relevant to configure, as the backend will just force quit I think if that's not configured to a longer timeout.

Reply:
That would be cool if it already works 👍, last time I played with that I was losing the connection to the database immediately once I hit "kubectl rollout restart deployment"

Reply:
Oh I see, that's a culprit, as we are actually using a proxy sidecar to DB

Reply:
Yeah you can configure those sidecars to wait I believe, or at least have a `preStop` command of `sleep` is a bit of hack to make sure it's the last thing to be shut down

Reply:
What's the sidecar you're using? We could just add support for the connector in knex like we have for cloud-sql-proxy too.

But ye, documenting this is also a good shout